File: frontend/src/types/analysis.ts
// ---------------------------------------------------------------------------
// Shared confidence types
// ---------------------------------------------------------------------------

/// Represents the different confidence tiers that can be assigned to findings and assertions,
/// based on their assessed reliability and credibility.
export type ConfidenceTier = "low" | "moderate" | "high" | "assessed";


/**
 * The confidence level of a particular finding, including its tier (e.g., low, moderate, high, assessed),
 * a numerical score, and the contributing factors such as authority, corroboration, independence, circularity,
 * and the types of sources that support it.
 */
export interface FindingConfidence {
  tier: ConfidenceTier;
  score: number;
  authority: number;
  corroboration: number;
  independence: number;
  circular_flag: boolean;
  source_types: string[];
}

/**
 * The confidence level of a particular assertion, including its tier (e.g., low, moderate, high, assessed),
 * a numerical score, and the contributing factors such as authority, corroboration, independence, circularity,
 * and the types of sources that support it.
 */
export interface AssertionConfidence {
  tier: ConfidenceTier;
  score: number;
  authority: number;
  corroboration: number;
  independence: number;
  circular_flag: boolean;
}

/**
 * Represents an assertion related to a particular perspective, including the assertion text, optional analysis,
 * the IDs of the findings that support this assertion, the types of sources that support it, and its confidence level.
 * This structure is used to organize the AI's key judgments and their supporting evidence in a way that is transparent
 * and traceable back to the underlying data.
 */
export interface PerspectiveAssertion {
  assertion: string;
  analysis?: string;
  supporting_finding_ids: string[];
  source_types: string[];
  confidence: AssertionConfidence | null;
}

// ---------------------------------------------------------------------------
// Collection Coverage (post-collection)
// ---------------------------------------------------------------------------

/**
 * A reference to a finding within the coverage analysis, including its ID, title, and source.
 */
export interface CoverageFindingRef {
  id: string;
  title: string;
  source: string;
}

/**
 * Represents the coverage score for a specific PIR question, including its index, text, priority level, confidence tier,
 * numerical score, the count of findings that address this PIR, the types of sources that contribute to this coverage,
 * a flag indicating whether there are any identified gaps in coverage, the rationale behind the score, and references to the specific findings.
 * This structure is used to provide a detailed breakdown of how well the collected information addresses each PIR question.
 */
export interface PirCoverageScore {
  pir_index: number;
  pir_question: string;
  priority: string;
  tier: ConfidenceTier;
  score: number;
  finding_count: number;
  source_types: string[];
  has_gap_flag: boolean;
  rationale: string;
  findings: CoverageFindingRef[];
}

/**
 * Represents the overall coverage analysis results after the collection phase, including the coverage scores for each PIR question,
 * the aggregate confidence tier and numerical score for the entire collection, and a summary of the coverage assessment.
 */
export interface CollectionCoverageResult {
  per_pir: PirCoverageScore[];
  aggregate_tier: ConfidenceTier;
  aggregate_score: number;
  summary: string;
}

// ---------------------------------------------------------------------------
// Processing & Analysis
// ---------------------------------------------------------------------------


export interface CouncilRunSettings {
  mode: "conference" | "quick";
  rounds: number;
  timeout_seconds: number;
  vote_retry_enabled: boolean;
  vote_retry_attempts: number;
}

/**
 * Represents a single finding extracted during the processing phase, including its ID, title, detailed finding text,
 * a summary of the evidence supporting it, the source of the finding, an assessed confidence score, the specific aspects it is relevant to,
 * any identified uncertainties, and an explanation of why this finding matters in the context of the overall analysis. Findings may also include
 * optional supporting data such as extracted entities, timestamps, locations, knowledge base references, attack IDs, domains, source URLs, source references, and indicators of compromise (IOCs).
 */
export interface ProcessingFinding {
  id: string;
  title: string;
  finding: string;
  evidence_summary: string;
  source: string;
  confidence: number;
  relevant_to: string[];
  supporting_data: Record<string, string[]>;
  why_it_matters: string;
  uncertainties: string[];
  computed_confidence: FindingConfidence | null;
}

export interface ProcessingResult {
  findings: ProcessingFinding[];
  gaps: string[];
}

/**
 * Represents the structured analysis generated during the analysis phase, including an overall title and summary,
 * a list of key judgments, a mapping of perspectives to their related assertions, a list of recommended actions based on the analysis,
 * and any identified information gaps that may require further investigation.
 */
export interface Analysis {
  title: string;
  summary: string;
  key_judgments: string[];
  per_perspective_implications: Record<string, PerspectiveAssertion[]>;
  recommended_actions: string[];
  information_gaps: string[];
}

/**
 * Represents the response from the analysis endpoint, including the processing results, the drafted analysis,
 * the latest council note if applicable,
 */
export interface CouncilTranscriptEntry {
  round: number;
  participant: string;
  response: string;
  timestamp: string;
  summary?: string;
}

export interface CouncilNote {
  status: string;
  question: string;
  participants: string[];
  rounds_completed: number;
  summary: string;
  key_agreements: string[];
  key_disagreements: string[];
  final_recommendation: string;
  full_debate: CouncilTranscriptEntry[];
  transcript_path: string | null;
}

export interface AnalysisResponse {
  processing_result: ProcessingResult;
  analysis_draft: Analysis;
  latest_council_note: CouncilNote | null;
  collection_coverage: CollectionCoverageResult | null;
  data_source: "session";
}

export interface RunAnalysisCouncilRequest {
  session_id: string;
  debate_point: string;
  finding_ids: string[];
  selected_perspectives: string[];
  council_settings?: CouncilRunSettings;
}

File: backend/src/models/analysis.py
"""Models for analysis-phase processing results, drafts, and council output."""

from pydantic import BaseModel, Field

from src.models.confidence import FindingConfidence, PerspectiveAssertion


class FindingModel(BaseModel):
    """A structured finding extracted during the processing phase."""

    id: str = Field(..., description="Unique finding identifier")
    title: str = Field(..., description="Short finding title")
    finding: str = Field(..., description="Detailed finding statement")
    evidence_summary: str = Field(..., description="Concise evidence summary")
    source: str = Field(..., description="Source category for the finding")
    confidence: int = Field(
        ..., ge=0, le=100, description="AI-generated confidence (legacy, 0–100)"
    )
    relevant_to: list[str] = Field(
        default_factory=list, description="Related PIR identifiers"
    )
    supporting_data: dict[str, list[str]] = Field(
        default_factory=dict, description="Structured support data for downstream use"
    )
    why_it_matters: str = Field(..., description="Analytical significance of finding")
    uncertainties: list[str] = Field(
        default_factory=list, description="Known uncertainties or unresolved questions"
    )
    computed_confidence: FindingConfidence | None = Field(
        default=None,
        description="Algorithm-computed confidence breakdown (None until enrichment pass)",
    )


class ProcessingResult(BaseModel):
    """Structured result from the processing phase."""

    findings: list[FindingModel] = Field(
        default_factory=list, description="Processed findings for analysis"
    )
    gaps: list[str] = Field(
        default_factory=list, description="Outstanding intelligence gaps"
    )


class AnalysisDraft(BaseModel):
    """Draft analysis produced from processed findings."""

    summary: str = Field(..., description="Short analytical summary")
    key_judgments: list[str] = Field(
        default_factory=list, description="Primary analytical judgments"
    )
    per_perspective_implications: dict[str, list[PerspectiveAssertion]] = Field(
        default_factory=dict,
        description="Implications grouped by analytical perspective, with evidence trace",
    )
    recommended_actions: list[str] = Field(
        default_factory=list, description="Recommended follow-up actions"
    )
    information_gaps: list[str] = Field(
        default_factory=list, description="Remaining information gaps"
    )


class CouncilTranscriptEntry(BaseModel):
    """A single council debate response."""

    round: int = Field(..., description="Round number")
    participant: str = Field(..., description="User-visible participant label")
    response: str = Field(..., description="Debate response text")
    timestamp: str = Field(..., description="ISO timestamp")


class CouncilRuntimeProfile(BaseModel):
    """Fixed application runtime profile for analysis-stage council runs."""

    adapter: str = Field(..., description="Council adapter/runtime family")
    model: str = Field(..., description="Default model identifier")
    mode: str = Field(..., description="Deliberation mode")
    rounds: int = Field(..., description="Configured round count")
    timeout_per_round_seconds: int = Field(
        ..., description="Configured timeout per round in seconds"
    )
    vote_retry_enabled: bool = Field(
        ..., description="Whether vote retry prompting is enabled"
    )
    vote_retry_attempts: int = Field(
        ..., description="Configured vote retry attempts"
    )
    working_directory: str = Field(..., description="Working directory for council")
    file_tree_injection_enabled: bool = Field(
        ..., description="Whether file-tree injection is enabled"
    )
    decision_graph_enabled: bool = Field(
        ..., description="Whether decision-graph context is enabled"
    )


class CouncilRunSettings(BaseModel):
    """User-configurable runtime overrides for analysis-stage council runs."""

    mode: str = Field(default="conference", pattern="^(conference|quick)$")
    rounds: int = Field(default=2, ge=1, le=5)
    timeout_seconds: int = Field(default=180, ge=30, le=900)
    vote_retry_enabled: bool = Field(default=True)
    vote_retry_attempts: int = Field(default=1, ge=0, le=3)


class CouncilNote(BaseModel):
    """Stable backend response model for analysis-stage council output."""

    status: str = Field(..., description="Council run status")
    question: str = Field(..., description="Debate question posed to the council")
    participants: list[str] = Field(
        default_factory=list, description="User-visible participant labels"
    )
    rounds_completed: int = Field(..., description="Number of rounds completed")
    summary: str = Field(..., description="Council summary / consensus text")
    key_agreements: list[str] = Field(
        default_factory=list, description="Key areas of agreement"
    )
    key_disagreements: list[str] = Field(
        default_factory=list, description="Key areas of disagreement"
    )
    final_recommendation: str = Field(
        ..., description="Final recommendation from the council"
    )
    full_debate: list[CouncilTranscriptEntry] = Field(
        default_factory=list, description="Structured debate transcript"
    )
    transcript_path: str | None = Field(
        default=None, description="Path to transcript file if available"
    )


class AnalysisSessionState(BaseModel):
    """Persisted analysis-stage state for one session."""

    session_id: str = Field(..., description="Session identifier")
    processing_result: ProcessingResult | None = Field(
        default=None, description="Persisted processing result"
    )
    analysis_draft: AnalysisDraft | None = Field(
        default=None, description="Persisted analysis draft"
    )
    latest_council_note: CouncilNote | None = Field(
        default=None, description="Most recent council note"
    )
